src/Orleans/Streams/Internal/StreamConsumerExtension.cs

```diff
@@ -149,15 +129,17 @@ public Task<StreamHandshakeToken> DeliverImmutable(GuidId subscriptionId, Stream
             }
             else
             {
-                // if no observer attached to the subscription, check if there's onSubscriptinChange actions defined
-                ISubscriptionChangeHandler handler;
-                if (this.onSubscriptionChangeActionMap.TryGetValue(batch.Value.GetType(), out handler))
+                // if no observer attached to the subscription, checkif the grain reference is IOnSubscriptionChangeActioner
+                if (this.onSubscriptionActioner != null)
                 {
                     //if the onAddAction attached an observer to the subscription
                     var streamProvider = this.providerRuntime.ServiceProvider
                                 .GetService<IStreamProviderManager>()
                                 .GetStreamProvider(streamId.ProviderName);
-                    await handler.InvokeOnAdd(streamId, subscriptionId, isRewindable, streamProvider);
+                   
+                    Type constructedType = typeof(OnSubscriptionActionInvoker<>).MakeGenericType(batch.Value.GetType());
```

Comment:
This is not safe.  In our test we tend to use heterogeneous streams, but most streams in the real world are streams of some event base type, with specific events being among a set of subclasses.  Like a stream of fruits, consisting of apples, oranges, and bananas.  This will create a stream of the first concrete event type, not the base type.

Two possible solutions I can foresee, though more may exist.
- Implement IOnSubscriptionActioner<T> for each type one wants to subscribe to.
- Untyped stream subscription handle, with a Resume<T>(..) call so users can specify the type they want to subscribe from.